src/Orleans/Runtime/OrleansThreadPool.cs

```diff
@@ -0,0 +1,1476 @@
+    {
+        private static readonly List<Thread> _workerThreads = new List<Thread>();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool QueueUserWorkItem(WaitCallback callBack, Object state)
+        {
+
+            return QueueUserWorkItemHelper(callBack, state);
+        }
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool QueueSystemWorkItem(WaitCallback callBack, Object state)
+        {
+
+            return QueueUserWorkItemHelper(callBack, state);
+        }
+
+        private static int gg;
+
+        private static readonly UnfairSemaphore _semaphore = new UnfairSemaphore();
+
+        public static void NotifyWorkItemComplete()
+        {
+            Interlocked.Decrement(ref gg);
+        }
+
+        public static void RequestWorkerThread()
+        {
+            _semaphore.Release();
+        }
+
+
+        static OrleansThreadPool()
+        {
+            _workerThreads.AddRange(Enumerable.Range(1, GrainClient.IsRunningInsideSilo ? Environment.ProcessorCount * 2 : 2)
+                .Select(v => new Thread(() =>
+                {
+                    while (true)
+                    {
+                        try
+                        {
+                            _semaphore.Wait();
+                            ThreadPoolWorkQueue.Dispatch();
+                        }
+                        catch (Exception ex)
+#if !NETSTANDARD
+                        when (!(ex is ThreadAbortException))
+#endif
+                        {
+                            // todo: normalize logging
+
+                            LogManager.GetLogger("OrleansThreadPool").Log(0, Severity.Error, "", null, ex);
+                       }
+                    }
+                })).ToList());
+
+            for (var i = 0; i < _workerThreads.Count; i++)
+            {
+                _workerThreads[i].Name = $"OrleansThreadPoolThread_{i.ToString()}";
+                _workerThreads[i].IsBackground = true;
+            }
+
+            _workerThreads.ForEach(v => v.Start());
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool QueueUserWorkItem(WaitCallback callBack)
+        {
+
+            return QueueUserWorkItemHelper(callBack, null);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool QueueSystemWorkItem(WaitCallback callBack)
+        {
+
+            return QueueUserWorkItemHelper(callBack, null);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool QueueUserWorkItemHelper(WaitCallback callBack, Object state)
+        {
+            bool success = false;
+
+            //
+            // If we are able to create the workitem, we need to get it in the queue without being interrupted
+            // by a ThreadAbortException.
+            //
+            try { }
+            finally
```

Comment:
It's done in finally block in order to avoid ThreadAbortException - http://stackoverflow.com/a/2186123/3725200